### PR TITLE
coll: TSP neighborhood gentran_linear algorithms

### PR DESCRIPTION
**Pull request by jain-surabhi-23**
Linear algorithms for neighborhood collectives rewritten using TSP based schedule.

### CODE REVIEW

**Comment by jain-surabhi-23** [comment]
@hzhou This is the PR which consolidates #3372 (alltoallw), #3375 (allgather), #3376 (allgatherv), #3377 (alltoall), #3378 (alltoallv). It addresses your review comments given before.

**Comment by hzhou** [comment]
test:jenkins/ch3/most
test:jenkins/ch4/most

**Comment by wesbland** [comment]
@jain-surabhi-23 Should we close the PRs listed above then?

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

File: src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c [hunk 1]
```diff
@@ -204,7 +207,47 @@ int MPIR_Ineighbor_alltoall_impl(const void *sendbuf, int sendcount,
...
     *request = NULL;
+    /* If the user picks one of the transport-enabled algorithms, branch there
+     * before going down to the MPIR_Sched-based algorithms. */
+    /* TODO - Eventually the intention is to replace all of the
+     * MPIR_Sched-based algorithms with transport-enabled algorithms, but that
+     * will require sufficient performance testing and replacement algorithms. */
+    if (comm_ptr->comm_kind == MPIR_COMM_KIND__INTRACOMM) {
+        /* intracommunicator */
+        switch (MPIR_Ineighbor_allgatherv_intra_algo_choice) {
+            case MPIR_INEIGHBOR_ALLTOALL_INTRA_ALGO_GENTRAN_LINEAR:
```
Comment (hzhou) [inline src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c]: Should the choice variable be `..._alltoall_...`?

File: src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c [hunk 2]
```diff
@@ -204,7 +207,47 @@ int MPIR_Ineighbor_alltoall_impl(const void *sendbuf, int sendcount,
...
+                goto fn_exit;
+                break;
+            default:
+                /* go down to the MPIR_Sched-based algorithms */
+                break;
+        }
+    } else {
+        /* intercommunicator */
+        switch (MPIR_Ineighbor_allgatherv_inter_algo_choice) {
+            case MPIR_INEIGHBOR_ALLTOALL_INTER_ALGO_GENTRAN_LINEAR:
```
Comment (hzhou) [inline src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c]: ditto

Comment (jain-surabhi-23) [reply src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c]: Fixed

File: src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c (hunk 1 above)
Comment (jain-surabhi-23) [reply src/mpi/coll/ineighbor_alltoall/ineighbor_alltoall.c]: Yes. I fixed it.

File: src/mpi/coll/src/coll_impl.c [hunk 3]
```diff
@@ -584,60 +584,88 @@ int MPII_Coll_init(void)
...
+    else if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLGATHERV_INTER_ALGORITHM, "gentran_linear"))
+        MPIR_Ineighbor_allgatherv_inter_algo_choice =
+            MPIR_INEIGHBOR_ALLGATHERV_INTER_ALGO_GENTRAN_LINEAR;
     else
         MPIR_Ineighbor_allgatherv_inter_algo_choice = MPIR_INEIGHBOR_ALLGATHERV_INTER_ALGO_AUTO;
 
     /* Ineighbor_alltoall Intra */
     if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALL_INTRA_ALGORITHM, "linear"))
         MPIR_Ineighbor_alltoall_intra_algo_choice = MPIR_INEIGHBOR_ALLTOALL_INTRA_ALGO_LINEAR;
+    else if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALL_INTRA_ALGORITHM, "gentra_linear"))
```
Comment (hzhou) [inline src/mpi/coll/src/coll_impl.c]: typo: `gentran_linear`

File: src/mpi/coll/src/coll_impl.c [hunk 4]
```diff
@@ -584,60 +584,88 @@ int MPII_Coll_init(void)
...
+        MPIR_Ineighbor_alltoall_intra_algo_choice =
+            MPIR_INEIGHBOR_ALLTOALL_INTRA_ALGO_GENTRAN_LINEAR;
     else
         MPIR_Ineighbor_alltoall_intra_algo_choice = MPIR_INEIGHBOR_ALLTOALL_INTRA_ALGO_AUTO;
 
     /* Ineighbor_alltoall Inter */
     if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALL_INTER_ALGORITHM, "linear"))
         MPIR_Ineighbor_alltoall_inter_algo_choice = MPIR_INEIGHBOR_ALLTOALL_INTER_ALGO_LINEAR;
+    else if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALL_INTER_ALGORITHM, "gentra_linear"))
```
Comment (hzhou) [inline src/mpi/coll/src/coll_impl.c]: ditto

File: src/mpi/coll/src/coll_impl.c (hunk 3 above)
Comment (jain-surabhi-23) [reply src/mpi/coll/src/coll_impl.c]: Fixed it. Sorry about that.

**Comment by hzhou** [comment]
test:jenkins/ch3/tcp

File: src/mpi/coll/src/coll_impl.c (hunk 3 above)
Comment (hzhou) [reply src/mpi/coll/src/coll_impl.c]: Oh no, nothing to be sorry about :) Typos are nasty. We need to do something (at some point) about those boilerplate infrastructures.

File: src/mpi/coll/src/coll_impl.c [hunk 5]
```diff
@@ -584,60 +584,88 @@ int MPII_Coll_init(void)
...
+        MPIR_Ineighbor_alltoall_inter_algo_choice =
+            MPIR_INEIGHBOR_ALLTOALL_INTER_ALGO_GENTRAN_LINEAR;
     else
         MPIR_Ineighbor_alltoall_inter_algo_choice = MPIR_INEIGHBOR_ALLTOALL_INTER_ALGO_AUTO;
 
     /* Ineighbor_alltoallv Intra */
     if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALLV_INTRA_ALGORITHM, "linear"))
         MPIR_Ineighbor_alltoallv_intra_algo_choice = MPIR_INEIGHBOR_ALLTOALLV_INTRA_ALGO_LINEAR;
+    else if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALLV_INTRA_ALGORITHM, "gentran_linear"))
+        MPIR_Ineighbor_alltoallv_intra_algo_choice = MPIR_INEIGHBOR_ALLTOALLV_INTRA_ALGO_LINEAR;
```
Comment (hzhou) [inline src/mpi/coll/src/coll_impl.c]: Not "..._ALGO_LINEAR", but "..._ALGO_GENTRAN_LINEAR"

File: src/mpi/coll/src/coll_impl.c [hunk 6]
```diff
@@ -584,60 +584,88 @@ int MPII_Coll_init(void)
...
     else
         MPIR_Ineighbor_alltoallv_intra_algo_choice = MPIR_INEIGHBOR_ALLTOALLV_INTRA_ALGO_AUTO;
 
     /* Ineighbor_alltoallv Inter */
     if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALLV_INTER_ALGORITHM, "linear"))
         MPIR_Ineighbor_alltoallv_inter_algo_choice = MPIR_INEIGHBOR_ALLTOALLV_INTER_ALGO_LINEAR;
+    else if (0 == strcmp(MPIR_CVAR_INEIGHBOR_ALLTOALLV_INTER_ALGORITHM, "gentran_linear"))
+        MPIR_Ineighbor_alltoallv_inter_algo_choice = MPIR_INEIGHBOR_ALLTOALLV_INTER_ALGO_LINEAR;
```
Comment (hzhou) [inline src/mpi/coll/src/coll_impl.c]: ditto.

File: src/mpi/coll/src/coll_impl.c (hunk 5 above)
Comment (jain-surabhi-23) [reply src/mpi/coll/src/coll_impl.c]: Fixed. Thanks for pointing it.

File: src/mpi/coll/src/coll_impl.c (hunk 6 above)
Comment (jain-surabhi-23) [reply src/mpi/coll/src/coll_impl.c]: Done

**Pull request merged**